### PR TITLE
executor, sessionctx, util: log a warning when memory usage of HashJoinExec exceed threshhold

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This PR does 2 things:
1. add a session variable called `tidb_exec_mem_thres` 
which is used to control the memory usage **warning threshold** 
in **Byte** of an executor during query execution.
When the memory usage hold by an executor 
exceeds the threshold, a warning log will be printed.

2. add a `updateMemoryUsage` interface for HashJoinExec.

PTAL @zz-jason @coocood

### CODE REVIEW

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -125,6 +129,7 @@ const (
 	DefCurretTS                   = 0
 	DefMaxChunkSize               = 1024
 	DefDMLBatchSize               = 20000
+	DefExecMemThres               = 4 * 1024 * 1024 * 1024 // 4 GiB
```
Comment (shenli) [inline sessionctx/variable/tidb_vars.go]: This is too small. 32 or 64 GB will be better.

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason 
``` go
func BenchmarkAtomic(b *testing.B) {
	var v int64 = 32*1024*1024*1024
	b.StartTimer()
	for i := 0; i < b.N; i ++ {
		a := atomic.LoadInt64(&v)
		atomic.StoreInt64(&v, a+int64(i))
	}
}

func BenchmarkSlice(b *testing.B) {
	sliceLen := 5
	v := make([]int64, sliceLen)
	for i := range v {
		v[i] = 32*1024*1024*1024
	}
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		sum := int64(0)
		for j := 0; j < sliceLen; j ++{
			sum += v[j]
		}
	}
}
```

```
Benchmark result:
BenchmarkAtomic-4                100000000               11.6 ns/op
BenchmarkSliceLen10-4          500000000                4.42 ns/op
BenchmarkSliceLen50-4          100000000               22.4 ns/op
BenchmarkSliceLen100-4          30000000                41.6 ns/op
```

**Comment by zz-jason** [comment]
@shenli @coocood PTAL

File: executor/join.go [hunk 2]
```diff
@@ -18,6 +18,7 @@ import (
 	"sync/atomic"
 	"unsafe"
 
+	"fmt"
```
Comment (coocood) [inline executor/join.go]: move it up

File: executor/join.go [hunk 3]
```diff
@@ -162,11 +165,30 @@ func (e *HashJoinExec) Open(goCtx goctx.Context) error {
...
 	return nil
 }
 
+func (e *HashJoinExec) updateMemoryUsage(idx int, memUsage int64) {
+	if e.memoryUsage[idx] >= memUsage {
+		return
+	}
+	e.memoryUsage[idx] = memUsage
+	var sum int64
+	for _, usage := range e.memoryUsage {
```
Comment (coocood) [inline executor/join.go]: This has data race.

File: executor/join.go [hunk 4]
```diff
@@ -729,6 +752,7 @@ func (e *HashJoinExec) join2Chunk(workerID int, outerChk *chunk.Chunk, joinResul
 			joinResult.err = errors.Trace(err)
 			return false, joinResult
 		}
+		e.updateMemoryUsage(workerID+1, joinResultChkBuffer.MemoryUsage())
```
Comment (coocood) [inline executor/join.go]: Can we only calculate the innerResult for now?
The join result is much smaller compared to innerResult.

File: executor/join.go [hunk 5]
```diff
@@ -162,11 +165,30 @@ func (e *HashJoinExec) Open(goCtx goctx.Context) error {
...
+	for _, usage := range e.memoryUsage {
+		sum += usage
+	}
+	memThreshold := e.ctx.GetSessionVars().ExecMemThreshold
+	if sum >= memThreshold {
+		detailedMsg := fmt.Sprintf("\n\t[inner buffer: %dB]", e.memoryUsage[0])
+		for i := 1; i < len(e.memoryUsage); i++ {
+			detailedMsg += fmt.Sprintf("\n\t[join worker %d: %dB]", i-1, e.memoryUsage[i])
+		}
+		log.Warnf(ErrMemExceedThreshold.GenByArgs("HashJoinExec", sum, memThreshold, detailedMsg).Error())
```
Comment (coocood) [inline executor/join.go]: It may log too many times.

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Comment by XuHuaiyu** [comment]
PTAL @coocood

File: sessionctx/variable/tidb_vars.go [hunk 6]
```diff
@@ -105,6 +105,10 @@ const (
 	// tidb_max_chunk_capacity is used to control the max chunk size during query execution.
 	TiDBMaxChunkSize = "tidb_max_chunk_size"
 
+	// tidb_exec_mem_threshold is used to control the memory usage warning threshold in Byte of an executor during query execution.
+	// When the memory usage hold by an executor exceeds the threshold, a warning log will be printed.
+	TiDBExecMemThres = "tidb_exec_mem_thres"
```
Comment (coocood) [inline sessionctx/variable/tidb_vars.go]: How about removing the `Exec`, most of the memories are consumed by executors.

**Comment by shenli** [comment]
@coocood PTAL

**Pull request merged**